src/Microsoft.AspNetCore.OData.Versioning/Extensions.DependencyInjection/MetadataControllerConfiguration.cs

```diff
@@ -0,0 +1,112 @@
+
+        ApiVersionConventionBuilder Conventions { get; }
+
+        public int Order => int.MaxValue;
+
+        public void OnProvidersExecuted( ApplicationModelProviderContext context )
+        {
+            var supported = new HashSet<ApiVersion>();
+            var deprecated = new HashSet<ApiVersion>();
+
+            DiscoverODataApiVersions( context, supported, deprecated, out var controllerModel );
+
+            if ( controllerModel == null )
+            {
+                return;
+            }
+
+            var builder = Conventions.Controller<VersionedMetadataController>()
+                                     .HasApiVersions( supported )
+                                     .HasDeprecatedApiVersions( deprecated );
+
+            builder.ApplyTo( controllerModel );
+        }
+
+        public void OnProvidersExecuting( ApplicationModelProviderContext context ) { }
+
+        void DiscoverODataApiVersions(
+            ApplicationModelProviderContext context,
+            HashSet<ApiVersion> supported,
+            HashSet<ApiVersion> deprecated,
+            out ControllerModel metadataControllerModel )
+        {
+            Contract.Requires( context != null );
+            Contract.Requires( supported != null );
+            Contract.Requires( deprecated != null );
+
+            metadataControllerModel = null;
+
+            var metadataController = typeof( VersionedMetadataController ).GetTypeInfo();
+
+            foreach ( var controllerModel in GetControllerModels( context ) )
+            {
+                var controller = controllerModel.ControllerType;
+
+                if ( metadataController.IsAssignableFrom( metadataController ) )
+                {
+                    metadataControllerModel = controllerModel;
+                }
+
+                var model = controllerModel.GetProperty<ApiVersionModel>();
+
+                if ( model == null )
+                {
+                    continue;
+                }
+
+                foreach ( var apiVersion in model.SupportedApiVersions )
+                {
+                    supported.Add( apiVersion );
+                }
+
+                foreach ( var apiVersion in model.DeprecatedApiVersions )
+                {
+                    deprecated.Add( apiVersion );
+                }
+            }
+
+            deprecated.ExceptWith( supported );
+        }
+
+        IEnumerable<ControllerModel> GetControllerModels( ApplicationModelProviderContext context )
+        {
+            Contract.Requires( context != null );
+            Contract.Ensures( Contract.Result<IEnumerable<ControllerModel>>() != null );
+
+            // TODO: is there a better way to do this? since we're already inside a pass of applying conventions, rebuild and apply
+            // the conventions without modifying the current pass to avoid unwanted side effects.
+            var provider = new DefaultApplicationModelProvider( Options );
```

Comment:
Could you inherit from DefaultApplicationModelProvider?

Reply:
I'm leery about extending or otherwise replacing any of the **IApplicationModelProviders**. Ideally, this code needs to run **after** all of the registered **IApplicationModelProviders** have run. I'm not entirely sure how to do that - yet. If I force it to run when it's not expected, I'm not sure what will happen; maybe nothing. The stop-gap approach here was to explicitly use **DefaultApplicationModelProvider** to get what's needed without changing the configuration until a better way is found.

Reply:
This makes sense